File: checks/ruby/lib/section508/checks/keyboard_operable.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Hover-only or click-only on a non-focusable node fails.
    class KeyboardOperable
      NATIVE = %w[BUTTON A INPUT SELECT TEXTAREA SUMMARY].freeze

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        hover_only = Dom.elements(root, "[onmouseover], [onmouseenter], [onmousemove]")
        hover_only.each do |element|
          return ["“#{hint(element)}” works on hover only."] unless keyboard_reachable?(element)
        end

        clicks = Dom.elements(root, "[onclick]")
        clicks.each do |element|
          next if native?(element) && (element.name.upcase != "A" || Dom.has_attr?(element, "href"))
          return ["“#{hint(element)}” is not reachable from the keyboard."] unless focusable?(element)
        end

        natives = Dom.elements(root, "button, a[href], input, select, textarea, summary")
        if hover_only.empty? && clicks.empty? && natives.empty?
          return ["This section has no keyboard-operable control."]
        end
        []
      end

      def self.keyboard_reachable?(element)
        return true if focusable?(element)

        element.css("button, a[href], input, select, textarea, summary, [tabindex]").any? do |node|
          next true if native?(node) && (node.name.upcase != "A" || Dom.has_attr?(node, "href"))

          node["tabindex"] && node["tabindex"] != "-1"
        end
      end

      def self.focusable?(element)
        return true if native?(element) && (element.name.upcase != "A" || Dom.has_attr?(element, "href"))

        tabindex = element["tabindex"]
        !tabindex.nil? && number(tabindex) >= 0
      end

      def self.native?(element)
        NATIVE.include?(element.name.upcase)
      end

      def self.hint(element)
        raw = element.text
        raw = nil if raw.empty?
        (raw || element["class"] || element.name.downcase).to_s.gsub(/\s+/, " ").strip
      end

      def self.number(value)
        Float(value)
      rescue ArgumentError, TypeError
        Float::NAN
      end
    end
  end
end

File: checks/python/src/section508/checks/keyboard_operable.py
from section508.dom import (
    SPACE,
    attr,
    class_string,
    has_attr,
    parse_number,
    select,
    tag_name,
    text_of,
)

NATIVE = {"button", "a", "input", "select", "textarea", "summary"}


def check(root):
    if not root:
        return []

    hover_only = select(root, "[onmouseover], [onmouseenter], [onmousemove]")
    for element in hover_only:
        if not is_keyboard_reachable(element):
            return [f"“{hint(element)}” works on hover only."]

    clicks = select(root, "[onclick]")
    for element in clicks:
        if tag_name(element) in NATIVE and (tag_name(element) != "a" or has_attr(element, "href")):
            continue
        if not is_focusable(element):
            return [f"“{hint(element)}” is not reachable from the keyboard."]

    natives = select(root, "button, a[href], input, select, textarea, summary")
    if not hover_only and not clicks and not natives:
        return ["This section has no keyboard-operable control."]
    return []


def is_keyboard_reachable(element) -> bool:
    if is_focusable(element):
        return True
    return bool(
        select(
            element,
            "button, a[href], input, select, textarea, summary, [tabindex]:not([tabindex='-1'])",
        )
    )


def is_focusable(element) -> bool:
    if tag_name(element) in NATIVE and (tag_name(element) != "a" or has_attr(element, "href")):
        return True
    tabindex = attr(element, "tabindex")
    number = parse_number(tabindex) if tabindex is not None else None
    return tabindex is not None and number is not None and number >= 0


def hint(element) -> str:
    return SPACE.sub(" ", text_of(element) or class_string(element) or tag_name(element)).strip()
